Add 'types' descriptor to the @view-transition rule.

mattwoodrow opened the pull request:
#### 7a2d4f4e88f0f51ac230dce6d25a977d675e50cf
<pre>
Add &apos;types&apos; descriptor to the @view-transition rule.
<a href="https://bugs.webkit.org/show_bug.cgi?id=278000">https://bugs.webkit.org/show_bug.cgi?id=278000</a>
&lt;<a href="https://rdar.apple.com/133727655">rdar://133727655</a>&gt;

Reviewed by Tim Nguyen.

@view-transition should support a &apos;types&apos; descriptor, that then get passed into the cross-document ViewTransition objects.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/at-rule-with-types-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/types-in-pagereveal-and-pageswap-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSViewTransitionRule.cpp:
(WebCore::StyleRuleViewTransition::StyleRuleViewTransition):
(WebCore::CSSViewTransitionRule::cssText const):
* Source/WebCore/css/CSSViewTransitionRule.h:
* Source/WebCore/css/CSSViewTransitionRule.idl:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeViewTransitionTypes):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveViewTransitionRule):
(WebCore::Document::reveal):
(WebCore::Document::dispatchPageswapEvent):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::resolveInboundCrossDocumentViewTransition):
(WebCore::ViewTransition::setTypes):
(WebCore::ViewTransition::createInbound): Deleted.
* Source/WebCore/dom/ViewTransition.h:

Canonical link: <a href="https://commits.webkit.org/282221@main">https://commits.webkit.org/282221@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/837dbdd3a88421b8f6f2d79587de7474c5e1730b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62494 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41849 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66478 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13046 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64614 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49536 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13382 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8992 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38884 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54128 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31105 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35596 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11448 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11974 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57216 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11764 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68207 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6440 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/57734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6469 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54161 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57931 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13882 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5352 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37649 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/38734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39831 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/38478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSViewTransitionRule.cpp:
```diff
@@ -54,6 +54,19 @@ StyleRuleViewTransition::StyleRuleViewTransition(Ref<StyleProperties>&& properti
+    if (auto value = properties->getPropertyCSSValue(CSSPropertyTypes)) {
+
+        auto processSingleValue = [&](const CSSValue& currentValue) {
+            if (currentValue.isCustomIdent())
```
Can it be anything else than a custom ident at this stage?

mattwoodrow replied:
It can be the not-custom 'none' ident.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8073,15 +8073,15 @@ void Document::dispatchPagehideEvent(PageshowEventPersistence persisted)
 
 // https://www.w3.org/TR/css-view-transitions-2/#vt-rule-algo
-bool Document::resolveViewTransitionRule()
+std::optional<Vector<String>> Document::resolveViewTransitionRule()
```
`std::optional<Vector<String>>` is a bit odd here. I wonder if we can better represent "skip transition" from the spec.

Maybe `std::pair<bool, Vector<String>>` or std::variant ? 

The code could read like this:
```
auto [skipTransition, types] = resolveViewTransitionRule();

or

switchOn(resolveViewTransitionRule(), ...)
```

nt1m replied:
Given the other uses of `resolveViewTransitionRule()` maybe `std::optional` is fine with a comment describing what the nullopt state is

mattwoodrow replied:
I don't like pair, since it means `{ true, { "hello" }}` is a syntactically valid return value, but has no semantic meaning.

I could try something like:

`class SkipTransition {};
std::variant<SkipTransition, Vector<String>>` if you prefer, maybe with a defined `operator bool` so that the other call sites remain simple.

nt1m replied:
Worth attempting your suggestion!

nt1m commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -183,6 +183,7 @@ class ViewTransition : public RefCounted<ViewTransition>, public CanMakeWeakPtr<
 
     const ViewTransitionTypeSet& types() const { return m_types; }
     void setTypes(Ref<ViewTransitionTypeSet>&& newTypes) { m_types = newTypes; }
+    void setTypes(const Vector<String>& newTypes);
```
Is ViewTransitionTypeSet reusable here?

tuankiet65 replied:
tl;dr: probably not, we'll have to create a new `ViewTransitionTypeSet` regardless.

As I understand, there's a one-way synchronization where if JS code manipulates `ViewTransition.types`, it'll actually manipulate a wrapper called `JSDOMSetLike` that maintains its own set content. Then `JSDOMSetLike` will forward the modification to `ViewTransitionTypeSet`, and it also maintains its own set content. But the opposite way doesn't exist - if the C++ code manipulates `ViewTransitionTypeSet`, we can't tell the `JSDOMSetLike` wrapper that the content of the set has changed, so now the C++ code and JS code will disagree on the content of the set.

tuankiet65 commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -93,20 +93,35 @@ Ref<ViewTransition> ViewTransition::createSamePage(Document& document, RefPtr<Vi
     viewTransition->m_updateCallbackDone.second->resolve();
     viewTransition->m_phase = ViewTransitionPhase::UpdateCallbackCalled;
 
+    viewTransition->setTypes(*types);
```
There's this `ViewTransition` constructor that sets the type so you don't have to write a separate routine to set the type:

```
ViewTransition::ViewTransition(Document& document, RefPtr<ViewTransitionUpdateCallback>&& updateCallback, Vector<AtomString>&& initialActiveTypes)
```

But I see this constructor assumes we have an update callback, which is not needed for multi-page transition. I think this is why you wrote a new constructor that doesn't take an update callback:

```
ViewTransition::ViewTransition(Document& document)
```
which creates an empty `ViewTransitionTypeSet` (my fault, I didn't understand the justification for the new constructor, so I just put empty)

I think we should do one of the following instead
* (preferred) extend the first constructor so that if the update callback is nullptr, then we don't call the callback
* extend the second constructor to take a vector of types

mattwoodrow replied:
The wording of the spec is that creating a `ViewTransition` and then setting the types on it are separate steps.

Your point about us needlessly creating an empty `ViewTransitionTypeSet`, only to replace it is a good one. I'll probably just pass the types in the constructor.

mattwoodrow replied:
I tried to extend the first constructor in the past, but we still need to differentiate between a single-page view transition with no callback (which still wants to fire the update promise), and a cross-doc view transition, which doesn't have a callback or use the promise.

tuankiet65 commented on Source/WebCore/css/CSSViewTransitionRule.h:
```diff
@@ -46,12 +46,14 @@ class StyleRuleViewTransition final : public StyleRuleBase {
 
     std::optional<ViewTransitionNavigation> navigation() const { return m_navigation; }
     ViewTransitionNavigation computedNavigation() const { return navigation().value_or(ViewTransitionNavigation::None); }
+    Vector<String> types() const { return m_types; }
```
Can we use `Vector<AtomString>` here too? So it'd match the `ViewTransitionTypeSet` side and we don't have to do conversions. Maybe there's a reason to use `String` here and I'd like to know. My original justification to use `AtomString` is because `AtomString` interns identical strings, and given that the type strings are identical in a lot of places, interning string could help reduce memory usage.

mattwoodrow replied:
I was using `Vector<String>` since that's the type generated by the IDL, but I just found the `[AtomString]` annotation that will hopefully fix that.

tuankiet65 commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -4799,5 +4799,24 @@ RefPtr<CSSPrimitiveValue> consumeAnchor(CSSParserTokenRange& range, CSSParserMod
+        auto type = consumeCustomIdent(range);
+        if (!type)
+            return nullptr;
+        if (type->customIdent().startsWith("-ua"_s))
```
The correct prefix is `-ua-` (reference: https://github.com/w3c/csswg-drafts/issues/9534#issuecomment-1802364085)

Anyway I think we should do this `-ua-` filtering in one place, namely within `ViewTransitionTypeSet`. This avoids having two identical filtering code (one here, one when matching `:active-view-transition-type()` selector). What do you think?

mattwoodrow replied:
That seems reasonable.

I'll have to create and store a `ViewTransitionTypeSet` inside `StyleRuleViewTransition`, and I can leave this subtest failing until you merge your PR.

mattwoodrow committed:
Add 'types' descriptor to the @view-transition rule.
https://bugs.webkit.org/show_bug.cgi?id=278000
<rdar://133727655>

Reviewed by Tim Nguyen.

@view-transition should support a 'types' descriptor, that then get passed into the cross-document ViewTransition objects.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/at-rule-with-types-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/types-in-pagereveal-and-pageswap-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSViewTransitionRule.cpp:
(WebCore::StyleRuleViewTransition::StyleRuleViewTransition):
(WebCore::CSSViewTransitionRule::cssText const):
* Source/WebCore/css/CSSViewTransitionRule.h:
* Source/WebCore/css/CSSViewTransitionRule.idl:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeViewTransitionTypes):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveViewTransitionRule):
(WebCore::Document::reveal):
(WebCore::Document::dispatchPageswapEvent):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::resolveInboundCrossDocumentViewTransition):
(WebCore::ViewTransition::setTypes):
(WebCore::ViewTransition::createInbound): Deleted.
* Source/WebCore/dom/ViewTransition.h:

Canonical link: https://commits.webkit.org/282221@main
